automation/balena-lib.inc

```diff
@@ -28,3 +46,97 @@ balena_lib_docker_pull_helper_image() {
 	fi
 	eval "$_retvalue"='${BALENA_YOCTO_SCRIPTS_REVISION}'
 }
+
+# Remove all versions of a  helper image
+# Inputs:
+# $1: Image name
+#
+balena_lib_docker_remove_helper_images() {
+    local _image_name="${1}"
+    local _image_ids
+    local _id
+    _image_ids=$(docker images "${NAMESPACE}/${_image_name}" --format "{{.ID}}")
+    for _id in ${_image_ids}; do
+        docker rmi -f "${_id}" || true
+    done
+}
+
+# Outputs to Balena environment according to Jenkins "deployTo" environment variables
+balena_lib_environment() {
+	local _api_env
+	if [ -n "${deployTo}" ]; then
+		if [ "${deployTo}" = "staging" ]; then
+			_api_env="balena-staging.com"
+		elif [ "${deployTo}" = "production" ]; then
+			_api_env="balena-cloud.com"
+		fi
+	else
+		echo "Defaulting to balena-cloud.com"
+		_api_env="balena-cloud.com"
+	fi
+	export BALENARC_BALENA_URL=${_api_env}
+	echo "${_api_env}"
+}
+
+# Output the Balena environment token according to set environment
+balena_lib_token() {
+	local _api_env
+	[ -z "${_api_env}" ] && _api_env=$(balena_lib_environment)
+	if [ "${_api_env}" = "balena-staging.com" ]; then
+		if [ -n "${BALENAOS_STAGING_TOKEN}" ]; then
+			_token="${BALENAOS_STAGING_TOKEN}"
+		fi
+	elif [ "${_api_env}" = "balena-cloud.com" ]; then
+		if [ -n "${BALENAOS_PRODUCTION_TOKEN}" ]; then
+			_token="${BALENAOS_PRODUCTION_TOKEN}"
+		fi
+	fi
+        if [ -z "${_token}" ]; then
+            if [ -f "${HOME}/.balena/token" ]; then
+                _token="$(cat "${HOME}/.balena/token")"
+            else
+                return
+            fi
+        fi
+	echo "${_token}"
+}
+
+# Logins to Balena cloud
+balena_lib_login() {
+	local _api_env=$1
+	local _token=$2
+
+	[ -z "${_api_env}" ] && _api_env=$(balena_enviroment)
+	if [ -z "${_api_env}" ]; then
+		echo "A Balena environment is required"
+		return
+	fi
+	[ -z "${_token}" ] && _token=$(balena_lib_token)
+	if [ -z "${_token}" ]; then
+		echo "A Balena token is required"
+		return
+	fi
+	BALENARC_BALENA_URL=${_api_env} balena login --token "${_token}"
+}
+
+# Login to dockerhub
+balena_lib_dockerhub_login() {
+    local _user="${1}"
+    local _pwd="${2}"
+
+    if [ -z "${_user}" ]; then
+        if [ -n "${JOB_NAME}" ]; then
```

Comment:
I can't find where `JOB_NAME` is actually used. Is it a legacy variable we should clean up?

Reply:
@klutchell it's set by Jenkins automatically. Here it is used to distinguish whether we are called as part of a Jenkins build so the credentials are expected to be in the environment, or not, in which case the login will ask the user for credentials (this is for local builds)

Reply:
Ah, of course, I forgot that we use JOB_NAME to determine if it's Jenkins.